proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+	Version string `json:"version,omitempty"`
+	// Channel specifies a channel that can be used to resolve a specific addon, eg: stable
+	// It will be ignored if Version is specified
+	Channel string `json:"channel,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+type PatchSpec struct {
+	Patches []*runtime.RawExtension `json:"patches,omitempty"`
+}
+```
+
+If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following
+
+```go
+// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate
+type NestedControlPlaneTemplateSpec struct {
+    // other fields ...    
+    
+    // NestedEtcd contains information required by the etcd provider to create the Etcd
+    NestedEtcd NestedComponentSpec `json:"nestedEtcd"`
+
+    // NestedKAS contains information required by the apiserver provider to create the apiserver
+    NestedKAS NestedComponentSpec `json:"nestedKAS"`
+
+    // NestedKCM contains information required by the controller-manager provider to create the controller-manager
+    NestedKCM NestedComponentSpec `json:"nestedKCM"`
+}
+```
+
+We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:
+
+```go
+type KASProvider string
+
+type EtcdProvider string
+
+type KCMProvider string
+
+const (
+    NativeKASProvider KASProvider = "NativeKASProvider"
+
+    NativeEtcdProvider  EtcdProvider = "NativeEtcdProvider"
+    EtcdClusterOperator EtcdProvider = "EtcdClusterOperator"
+    EtcdOperator        EtcdProvider = "EtcdOperator"
+
+    NativeKCMProvider KCMProvider = "NativeKCMProvider"
+)
+```
+
+We will add three new fields `NestedEtcdProvider`, `NestedKASProvider`, and the `NestedKCMProvider` into the NCP, which will make the NCP look like the following:
+
+```go
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+    // other fields.... 
+    
+    // NestedEtcdProvider specifies which EtcdProvider will be used to create the Etcd
+    // +optional
+    NestedEtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
+     
+    // NestedKASProvider specifies which KASProvider will be used to create the kube-apiserver
+    // +optional
+    NestedKASProvider KASProvider `json:"kasProvider,omitempty"`
+ 
+    // NestedKCMProvider specifies which KCMProvider will be used to create the kube-controller-manager
+    // +optional
+    NestedKCMProvider KCMProvider `json:"kcmProvider,omitempty"`
+}
```

Comment:
```suggestion
type NestedControlPlaneSpec struct {
    // other fields.... 
    
    // EtcdProvider specifies which EtcdProvider will be used to create the Etcd
    // +optional
    EtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
     
    // ApiServerProvider specifies which KASProvider will be used to create the kube-apiserver
    // +optional
    ApiServerProvider KASProvider `json:"apiServerProvider,omitempty"`
 
    // ControllerManagerProvider specifies which KCMProvider will be used to create the kube-controller-manager
    // +optional
    ControllerManagerProvider KCMProvider `json:"controllerManagerProvider,omitempty"`
}
```

Same suggestion as above regarding naming.

On a more general note, do we need pluggable providers for the first iteration? Would it be ok to reduce scope a bit and only allow the built-in controllers for etcd/api-server/controller-manager instead?

Reply:
We don't fully as it's documented IMO, this was called out a couple places. Specifically here - https://github.com/kubernetes-sigs/cluster-api-provider-nested/pull/11/files/4b6d8aea98ed04eea543f2e29319567d4bab5207#r521166698 As long as we have a way to swap the controllers for specific implementations that would be sufficient, IE I could use an etcd prefix or an operator backing the `NEtcd` controller instead and manage that out-of-tree.

I'm very hesitant to add too much pluggability upfront except controlling what controller manages each object. Since there are so much prior art in the OSS for etcd controllers and we should be able to support those without forking.